chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,378 @@
+
+        return prediction, labels
+
+    def __call__(self, args):
+        """Applies discriminative margin based clustering loss
+
+        Steps are:
+            - Reshape inputs to prepare for loss calculation
+            - Calculate means
+            - Calculate variance term
+            - Calculate distance term
+            - Calculate regularization term
+            - Add weights to all and return loss value
+
+        Args:
+            x (tuple) : Contains several inputs
+                - x[0] = segmentation prediction output
+                - x[1] = segmentation ground truth
+                - x[2] = number of objects in ground truth
+                - x[3] = indexes of non-zero ground truths
+        Returns:
+            float : Loss value
+        """
+
+        buffer = chainer.config.type_check
+        chainer.config.type_check = False
+
+        assert(len(args) == 4)
+
+        # Inputs
+        prediction, labels = self._prepare_inputs(args[0], args[1])
+        n_objects, gt_idx = args[2:4]
+
+        # Calculate cluster means
+        c_means = self._means(prediction, labels, n_objects,
+                              self.max_n_clusters, gt_idx)
+
+        # Calculate losses
+        l_var = self._variance_term(prediction, labels, c_means, self.delta_v,
+                                    gt_idx)
+        l_dist = self._distance_term(c_means, self.delta_d, n_objects)
+        l_reg = self._regularization_term(c_means, n_objects)
+
+        chainer.config.type_check = buffer
+        return self.alpha * l_var + self.beta * l_dist + self.gamma * l_reg
+
+
+def discriminative_margin_based_clustering_loss(
+        x, delta_v, delta_d, max_n_clusters,
+        norm=1, alpha=1.0, beta=1.0, gamma=0.001):
+    """Discriminative margin based clustering loss function
+
+    This is the implementation of following paper:
+    https://arxiv.org/pdf/1802.05591.pdf
+
+    In segmentation, one of the biggest problem is having noise at the output
+    of a trained network.
+    For cross-entropy based approaches, if the pixel value is wrong,
+    the loss value will be same independent from wrong pixel's location.
+    However, for segmentation, even though network gives wrong pixel output,
+    it is desirable to have it
+    as close as possible to the original position.
+    By applying discriminative loss function, groups of segmentation instances
+    can be moved together.
+
+    This loss function calculates three different parameters:
+        - Variance Loss:
+            Loss to penalize distances between pixels which are belonging
+            to same instance. (Pull force)
+        - Distance loss:
+            Loss to penalize distances between the centers of instances.
+            (Push force)
+        - Regularization loss:
+            Small regularization loss to penalize weights against overfit
+
+    Args:
+        x (tuple) : Contains several inputs
```

Comment:
Could you explicitly write that each elements of `x` is either `numpy.ndarray`, `cupy.ndarray` or `Variable` holding them (e.g. [F.ReLU](https://github.com/chainer/chainer/blob/master/chainer/functions/activation/relu.py#L152))